pkg/compact/downsample/downsample.go

```diff
@@ -0,0 +1,585 @@
+		if j > len(chunked) {
+			j = len(chunked)
+		}
+		length := 0
+		for _, p := range chunked[:j] {
+			length += len(p)
+		}
+		part := data[:length]
+		data = data[length:]
+		chunked = chunked[j:]
+
+		// Downsample the counter series spanning the chunk batch into a single chunk.
+		downsampleCounter(ser, part, window, false)
+	}
+}
+
+func downsampleCount(ser *series, data []sample, window int64) {
+	nextT := int64(-1)
+	var count int
+
+	for _, s := range data {
+		if s.t > nextT {
+			if nextT != -1 {
+				ser.add(nextT, float64(count))
+			}
+			count = 0
+			nextT = s.t - (s.t % window) + window - 1
+		}
+		count++
+	}
+	ser.add(nextT, float64(count))
+}
+
+func downsampleSum(ser *series, data []sample, window int64) {
+	nextT := int64(-1)
+	var sum float64
+
+	for _, s := range data {
+		if s.t > nextT {
+			if nextT != -1 {
+				ser.add(nextT, sum)
+			}
+			sum = 0
+			nextT = s.t - (s.t % window) + window - 1
+		}
+		sum += s.v
+	}
+	ser.add(nextT, sum)
+}
+
+func downsampleMin(ser *series, data []sample, window int64) {
+	nextT := int64(-1)
+	min := math.MaxFloat64
+
+	for _, s := range data {
+		if s.t > nextT {
+			if nextT != -1 {
+				ser.add(nextT, min)
+			}
+			min = math.MaxFloat64
+			nextT = s.t - (s.t % window) + window - 1
+		}
+		if s.v < min {
+			min = s.v
+		}
+	}
+	ser.add(nextT, min)
+}
+
+func downsampleMax(ser *series, data []sample, window int64) {
+	nextT := int64(-1)
+	max := -math.MaxFloat64
+
+	for _, s := range data {
+		if s.t > nextT {
+			if nextT != -1 {
+				ser.add(nextT, max)
+			}
+			max = -math.MaxFloat64
+			nextT = s.t - (s.t % window) + window - 1
+		}
+		if s.v > max {
+			max = s.v
+		}
+	}
+	ser.add(nextT, max)
+}
+
+// countChunkSeriesIterator iterates over an ordered sequence of chunks and treats decreasing
+// values as counter reset.
+// Additionally, it can deal with downsampled counter chunks, which set the last value of a chunk
+// to the original last value. The last value can be detected by checking whether the timestamp
+// did not increase w.r.t to the previous sample
+type countChunkSeriesIterator struct {
+	chks   []chunkenc.Iterator
+	i      int     // current chunk
+	total  int     // total number of processed samples
+	lastT  int64   // timestamp of the last sample
+	lastV  float64 // value of the last sample
+	totalV float64 // total counter state since beginning of series
+}
+
+func (it *countChunkSeriesIterator) LastSample() (int64, float64) {
+	return it.lastT, it.lastV
+}
+
+func (it *countChunkSeriesIterator) Next() bool {
+	if it.i >= len(it.chks) {
+		return false
+	}
+	// Chunk ends without special sample. It was not a downsampled counter series.
+	if ok := it.chks[it.i].Next(); !ok {
+		it.i++
+		return it.Next()
+	}
+	t, v := it.chks[it.i].At()
+	// First sample sets the initial counter state.
+	if it.total == 0 {
+		it.total++
+		it.lastT, it.lastV = t, v
+		it.totalV = v
+		return true
+	}
+	// If the timestamp increased, it is not the special last sample.
```

Comment:
I don't get this. In `downsampleCounter` we are adding special sample as `nextT` so... increased timestamp

Reply:
I think it is not true for case
```
if split && i > 0 && i%n == 0 {
 +				// Add signaling sample that indicates what the true last value was
 +				// before cutting a new chunk.
 +				// For already downsampled series lastV is set to the last value that
 +				// series itself added as a signaling sample.
 +				ser.add(nextT, lastV)
 +				ser.cut()
 +			}
```

Reply:
Right, because adding of the counter sample was actually done after the block you showed. They have to be swapped.